### DISCUSSION TITLE
Developer's Guide

### DISCUSSION THREAD

**Discussion by bptree88** [mucommander/mucommander#1133] [General]
Could anyone tell me the link to the guide?  
I'm trying to set up a project in Eclipse 2023, but get 10K plus compile errors, eg., About unresolved type, AbstractAction unresolved type, etc.  What am I missing?  Thanks.
My system: Win 11 Home, Eclipse 2023, JDK 21.0.1.

**Comment by pskowronek** [comment]
For Intellij Idea I've just built the project from command line and then imported it to Idea w/o any issues.
Try building the project (`./gradlew clean run`) first, if it succeeds then try to import it to Eclipse.
You may want to execute `git submodule update --remote` (`jre` is done as submodule).

If it succeeds, then you may want to update this wiki page: https://github.com/mucommander/mucommander/wiki/Developer-Resources to help others. 

However, if you have 10k+ errors, then probably something is terribly wrong with your Eclipse configuration/setup. Just guessing.

**Comment by ahadas** [comment]
@bptree88 did you import the project as a Gradle project?

**Comment by infinitesoft13** [comment]
I create new project from url - intellij.
Use this link: https://github.com/mucommander/mucommander.git
I have this error 
![errorIntelliJ](https://github.com/mucommander/mucommander/assets/72708474/81753bbd-026b-45ac-8e54-a91f58c17bfe)

debug the program i have this error:
Caused by: java.lang.ClassNotFoundException: com.beust.jcommander.JCommander

Thanks
Francesco

**Reply by pskowronek** [reply]
> https://github.com/mucommander/mucommander.git

try `git clone https://github.com/mucommander/mucommander.git` from terminal. Then you need to init submodule (`jre` is submodule): https://stackoverflow.com/a/1032653/1715521. If this works, then import the project to Idea.

**Comment by infinitesoft13** [comment]
Thanks!!!
try:  
git clone https://github.com/mucommander/mucommander.git
git submodule update --init --remote --recursive
git submodule init      
 git submodule update --recursive    

Open project on intellij is ok!!!
Debug muCommander in scr/Main/java package com.mucommander.mai
I have this error:
Caused by: java.lang.ClassNotFoundException: com.beust.jcommander.JCommander

Thanks a lot
Francesco

**Reply by pskowronek** [reply]
> Caused by: java.lang.ClassNotFoundException: com.beust.jcommander.JCommander

You wanted to run/debug the main directly via Idea? You cannot do this just like that, this is OSGi project. You have to run/debug it via `gradle run` (and set env `DEBUG=5001`), then attach debugger to it. Idea smartly understands gradle (somehow :P) and it provides such options. Personally I run `./gradlew clean run` from terminal and then attach debugger to it (for me it is kinda quicker).

**Comment by infinitesoft13** [comment]
Yes: I wanted to debug directly via Idea!!!
Sorry but I didn't understand how to debug with gradle?! Can you point me to a link with explanations?
Thanks a lot
Francesco

**Comment by pskowronek** [comment]
Terminal:
`export DEBUG=5001; ./gradlew clean run`
then in Idea: Run -> Debug 'remote debug' or 'Debug...' (and point idea to localhost and 5001 port), or 'Attach to process' .

In Idea set breakpoint to the place of interest (for example in main). pls read about remote java debugging.

**Reply by pskowronek** [reply]
alternatively, instead of using Terminal (via terminal it builds faster for me), you can try: Idea -> Debug... and there:
![obraz](https://github.com/mucommander/mucommander/assets/360665/b3d62947-224b-4a28-9beb-b3107a78f428)

then you need to attach to the port.

**Comment by infinitesoft13** [comment]
1. execute export DEBUG=5001; ./gradlew clean run
Where is remote debug? Use Run - Attach To process but not start...
![AttachProcess](https://github.com/mucommander/mucommander/assets/72708474/4da64677-64c7-4668-9a80-a73fb73278cd)

2.  on debug IDE import Grade muCommander
![debugGradleMuCommander](https://github.com/mucommander/mucommander/assets/72708474/ff7ddcc6-8b2a-418d-8850-a9c866186d42)
Debug muCommander gradle and have this output:

8:05:45: Executing...

Starting Gradle Daemon...
Connected to the target VM, address: '127.0.0.1:59931', transport: 'socket'
Gradle Daemon started in 539 ms

> Task :help

Welcome to Gradle 8.7.

To run a build, run gradle <task> ...

To see a list of available tasks, run gradle tasks

To see more detail about a task, run gradle help --task <task>

To see a list of command-line options, run gradle --help

For more detail on using Gradle, see https://docs.gradle.org/8.7/userguide/command_line_interface.html

For troubleshooting, visit https://help.gradle.org

Deprecated Gradle features were used in this build, making it incompatible with Gradle 9.0.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/8.7/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 5s
1 actionable task: 1 executed
08:05:50: Execution finished.
Disconnected from the target VM, address: '127.0.0.1:59931', transport: 'socket'

Thanks a lot

**Reply by pskowronek** [reply]
> 1. execute export DEBUG=5001; ./gradlew clean run
>        Where is remote debug? Use Run - Attach To process but not start...

Switch to traditional UI.
https://www.jetbrains.com/help/idea/new-ui.html

> 
>     2. on debug IDE import Grade muCommander

specify task to run, like `run` or `runOsgi`


> To see a list of available tasks, run gradle tasks

Read the output - try to understand it. Above you got suggestion who you can see gradle tasks by ...... running `gradle tasks`.

> 
> To see more detail about a task, run gradle help --task
The same here.

> 
> To see a list of command-line options, run gradle --help

And here.

> 
> For more detail on using Gradle, see https://docs.gradle.org/8.7/userguide/command_line_interface.html

Worth to read that page.

> 
> For troubleshooting, visit https://help.gradle.org

Or this one.

Read how to debug java apps with idea: https://www.jetbrains.com/help/idea/debugging-your-first-java-application.html

I really recommend running build from terminal/shell and then attach Idea debuger to it - you could understand how it works more easily I think.

**Reply by infinitesoft13** [reply]
Change a traditional UI but not view remote debug...

Thank you so much... I'll try to understand
Francesco

**Reply by pskowronek** [reply]
![obraz](https://github.com/mucommander/mucommander/assets/360665/b6e1e20f-cda6-4215-b387-1118bfc3c123)

![obraz](https://github.com/mucommander/mucommander/assets/360665/ed8d31dc-0569-4b6f-b450-1a8753acddb8)

**Reply by pskowronek** [reply]
> Change a traditional UI but not view remote debug...
> 
> Thank you so much... I'll try to understand Francesco

https://github.com/mucommander/mucommander/wiki/Debug-from-IntelliJ

**Reply by infinitesoft13** [reply]
Sorry but I wasn't able to test anymore (I had problems with the house, it's raining and water gets in...) also I have to leave... as soon as I get back and I have a moment I'll try!!!
Thank you again!!!
